cni/ipam/ipam_test.go

```diff
@@ -27,70 +30,191 @@ var ipamQueryResponse = "" +
-		return
-	}
+func getIfName() (string, error) {
 
-	// Configure test mode.
-	plugin.SetOption(common.OptEnvironment, common.OptEnvironmentAzure)
-	plugin.SetOption(common.OptAPIServerURL, "null")
-	plugin.SetOption(common.OptIpamQueryUrl, "http://"+ipamQueryUrl)
+	ifName := ""
 
-	// Start the plugin.
-	err = plugin.Start(&config)
+	interfaces, err := net.Interfaces()
 	if err != nil {
-		fmt.Printf("Failed to start IPAM plugin, err:%v.\n", err)
-		return
+		return ifName, err
 	}
 
-	// Run tests.
-	exitCode := m.Run()
-
-	// Cleanup.
-	plugin.Stop()
-	testAgent.Stop()
+	for _, iface := range interfaces {
+		ifName = iface.Name
+		break
+	}
 
-	os.Exit(exitCode)
+	return ifName, err
 }
 
-// Handles queries from IPAM source.
-func handleIpamQuery(w http.ResponseWriter, r *http.Request) {
-	w.Write([]byte(ipamQueryResponse))
-}
-
-//
-// CNI IPAM API compliance tests
-// https://github.com/containernetworking/cni/blob/master/SPEC.md
-//
-
-func TestAddSuccess(t *testing.T) {
-}
 
-func TestDelSuccess(t *testing.T) {
-}
+var (
+	_ = Describe("Test IPAM", func() {
+
+		var (
+			config common.PluginConfig
+			plugin *ipamPlugin
+			testAgent *common.Listener
+			arg *cniSkel.CmdArgs
+			ifName string
+			err error
+		)
+
+		BeforeSuite(func() {
+			// Create a fake local agent to handle requests from IPAM plugin.
+			u, _ := url.Parse("tcp://" + ipamQueryUrl)
+			testAgent, err = common.NewListener(u)
+			Expect(err).NotTo(HaveOccurred())
+
+			testAgent.AddHandler("/", handleIpamQuery)
+
+			err = testAgent.Start(make(chan error, 1))
+			Expect(err).NotTo(HaveOccurred())
+
+			ifName, err = getIfName()
+			Expect(err).NotTo(HaveOccurred())
+
+			arg = &cniSkel.CmdArgs{IfName:ifName}
+		})
+
+		AfterSuite(func() {
+			// Cleanup.
+			plugin.Stop()
+			testAgent.Stop()
+		})
+
+		Context("IPAM start", func() {
+			It("Create IPAM plugin", func() {
+				// Create the plugin.
+				plugin, err = NewPlugin("ipamtest", &config)
+				Expect(err).NotTo(HaveOccurred())
+			})
+
+			It("Start IPAM plugin", func() {
+				// Configure test mode.
+				plugin.SetOption(common.OptEnvironment, common.OptEnvironmentAzure)
+				plugin.SetOption(common.OptAPIServerURL, "null")
+				plugin.SetOption(common.OptIpamQueryUrl, "http://"+ipamQueryUrl)
+				// Start the plugin.
+				err = plugin.Start(&config)
+				Expect(err).NotTo(HaveOccurred())
+			})
+		})
+
+		Describe("Test IPAM ADD and DELETE pool", func() {
+
+			var result *cniTypesCurr.Result
+
+			Context("When ADD with nothing (request pool)", func() {
+				It("Request pool and ADD successfully", func() {
+					arg.StdinData = getStdinData("0.4.0", ifName, "", "")
+					err = plugin.Add(arg)
+					Expect(err).ShouldNot(HaveOccurred())
+					result, err = parseResult(arg.StdinData)
+					Expect(err).ShouldNot(HaveOccurred())
+					address1, _ := platform.ConvertStringToIPNet("10.0.0.5/16")
+					address2, _ := platform.ConvertStringToIPNet("10.0.0.6/16")
```

Comment:
do we need address2 here?

Reply:
I think may yes, while testing, plugin.Add(arg) would return 10.0.0.5 or 10.0.0.6 randomly.

Reply:
@tamilmani1989 , should this be random behaviour? I'd assume it'll always be 10.0.0.5 from a clean cluster

Reply:
In Golang, traversal map does not seem to be sorted by key, so when traversing Addresses map [string]*addressRecord, it may not necessarily take out 10.0.0.5?
````go
	// If no address was found, return any available address.
	if ar == nil {
		for _, ar = range ap.Addresses {
			if !ar.InUse && ar.ID == "" {
				break
			}
			ar = nil
		}

		if ar == nil {
			return "", errNoAvailableAddresses
		}
	}
```

Reply:
@jaer-tsun  yes it can return in any order

Reply:
Sorry, my fault, please keep your code as it is.